rfd/0207-encrypted-session-recordings.md

```diff
@@ -0,0 +1,284 @@
+  ENCRYPTED_FILE_RECIPIENT_TYPE_SSH_ED25519 = 2;
+  // RSA public key marshaled as an ssh-rsa public key.
+  // RSA keys must have at least 2048 bits.
+  ENCRYPTED_FILE_RECIPIENT_TYPE_SSH_RSA = 3;
+  // Ed25519 public key marshaled in PKIX PEM form.
+  ENCRYPTED_FILE_RECIPIENT_TYPE_ED25519 = 4;
+  // RSA public key marshaled in PKIX PEM form.
+  // RSA keys must have at least 2048 bits.
+  ENCRYPTED_FILE_RECIPIENT_TYPE_RSA = 5;
+}
+
+message EncryptedFileRecipient {
+  // Recipient public key type.
+  // * age_x25519
+  // * ssh_ed25519 
+  // * ssh_rsa
+  // * ed25519
+  // * rsa
+  string recipient_type = 1 [(gogoproto.jsontag) = "recipient_type"];
+  // Public key marshaled according to recipient_type.
+  string public_key = 2 [(gogoproto.jsontag) = "public_key"];
+}
+```
+
+### Key Generation
+
+Admins are free to configure their own public key recipients provided they are
+generated by a supported algorithm.
+
+As a convenience, we will also add a new `keygen` utility to `tsh` for key
+generation. By default, this will generate a new X25519 key pair to be
+configured as a recipient, and it will also provide integration with an HSM through a
+set of `--hsm*` parameters. More information about this in
+[HSM Configuration](#hsm-configuration).
+
+### Session Recording Modes
+
+There are four session recording modes that describe where the recordings are
+captured and how they're shipped to long term storage.
+- `proxy-sync`
+- `proxy`
+- `node-sync`
+- `node`
+Where the recordings are collected is largely unimportant to the encryption strategy,
+but whether or not they are handled async or sync has different requirements.
+
+In sync modes the session recording data is written immediately to long term storage
+without intermediate disk writes. This means we can simply instantiate an age
+encryptor at the start of the session and encrypt the recording data as it's sent
+to long term storage.
+
+In async modes the session recording data is written to intermediate `.part` files.
+These files are collected until they're ready for upload and are then combined into
+a single `.tar` file. In order to encrypt individual parts, we will build a special
+`io.Writer` that contains a single instance of the age encryptor that can proxy
+writes across multiple files. This will require maintaining a concurrency-safe
+mapping of in-progress uploads to encrypted writers which incurs a bit of added
+complexity. However it intentionally avoids any sort of intermediate key management
+which seems a worthwhile tradeoff.
+
+### Protocols
+
+We record sessions for multiple protocols, including ssh, k8s, database sessions and
+more. Because this approach encrypts at the point of writing without modifying the
+recording structure, the strategy for encryption is expected to be the same across
+all protocols.
+
+### Decryption and Replay
+
+Replay of encrypted session recordings will happen through `tsh play`. For
```

Comment:
What will we do about desktop session recordings, which can only be played via the UI?

Reply:
With the current design they wouldn't be replayable. In theory we could download and decrypt all of the PNGs with a command, but we would have to stitch the chunks together into full PNGs and the UX would be pretty poor.

I had an idea to add something like a `tsh decrypt-server` to support replays on web. It would spin up a local server capable of unwrapping the decryption key at an address that could be pasted into the web UI. The web UI would send encrypted recording data in and get decrypted recording data out, all local to the user's machine. I've seen a similar setup work before, but I haven't fully fleshed it out yet.

Reply:
It's definitely an option - I have always thought of doing a `teleport play desktop` or `tsh play desktop` that would spin up a local webserver that can play a desktop session for offline playback, though this might be a big effort.

Can auth decrypt the recordings on the fly? Recordings play through auth's `StreamSessionEvents` API, so if auth has access to the HSM can it stream decrypted events to the client?